linalg/sparsemat.hpp

```diff
@@ -22,6 +22,20 @@
 #include "../general/globals.hpp"
 #include "densemat.hpp"
 
+#if defined(MFEM_USE_HIP)
+#include <hipsparse.h>
+#define cu_or_hip(stub) hip##stub
+#define Cu_or_Hip(stub) Hip##stub
+#define CU_or_HIP(stub) HIP##stub
+#define CUDA_or_HIP(stub) HIP##stub
+#elif defined(MFEM_USE_CUDA)
+#define cu_or_hip(stub) cu##stub
+#define Cu_or_Hip(stub) Cu##stub
+#define CU_or_HIP(stub) CU##stub
+#define CUDA_or_HIP(stub) CUDA##stub
+#endif
```

Comment:
1. This is probably not the best place to define these, how about `device.hpp`?

2. Any macro we define should have an `MFEM` prefix.

3. To keep it short, maybe we can name them `MFEM_cu` or `MFEM_gpu`?

Reply:
> This is probably not the best place to define these, how about `device.hpp`?
> Any macro we define should have an `MFEM` prefix.
> To keep it short, maybe we can name them `MFEM_cu` or `MFEM_gpu`?

In my opinion the names will be hard to understand if they are shorter and I think that they should probably be called something like `MFEM_SUB_{a}_or_{b}` (I realize this is longer but I think it's worth it for readability). With the shorter names we'd have a macro like `MFEM_CUDA_or_HIP` that isn't what people (or at least I) would expect.
I also wasn't sure what the general feedback on the macro assisted typenames would be but if all this sounds okay I will rename to `MFEM_SUB...` and move them to `device.hpp`.

Reply:
If the only place these macros are used is in `sparsemat.cpp` -- that's where they should be. There is no reason to make them public by adding them to a public header. If the macros are used in the header `sparsemat.hpp`, just move that code to the `.cpp` file.

Reply:
@v-dobrev , I like that idea, I'll make the changes

Reply:
thanks @tomstitt

Reply:
done